### ISSUE TITLE
Comparison of reflect.Value types

### ISSUE THREAD

**Issue by stewi1014** [maxatome/go-testdeep#107]
Here to annoy you again :P

I'm aiming to support the encoding of reflect.Type and reflect.Value in my encoding library.
go-testdeep has been amazing for tests, but test cases where I compare a reflect.Value values, go-testdeep compares them like a typical struct, and doesn't compare the underlying value inside reflect.Value.

This means a reflect.Value of a value is only equal if the underlying value is at the same pointer address, and the reflect.Value was created in the same way due to how reflect uses the flag field in reflect.Value.

In any case, I don't think the semantics of how reflect.Value works is important. For my use case, I'm looking to compare the value inside reflect.Value, and I don't care so much about the reflect.Value itself, e.g if both are addressable or if only one is, although an argument could be made for this. I believe something like 
```go
if got.Type() == reflect.TypeOf(reflect.Value{}) {
    return deepValueEqual(ctx, got.Interface().(reflect.Value), expected.Interface().(reflect.Value))
}
```
in deepValueEqual could work, but I'm unsure how this might work with the rest of the codebase. I also know that the ptr field of reflect.Value could introduce an infinite recursion, an extreme case being a reflect.Value that is the value of itself, but less contrived examples do exist.

I'd be happy to experiment implementing this and submit a PR. Is this something you'd want to support?

**Comment by maxatome** [comment]
Hi,
So you want `reflect.Value` be handled specifically so that only their content should be compared, and not the reflect.Value itself.
```go
a, b := reflect.ValueOf(1234), reflect.ValueOf(1234)
td.Cmp(t, a, a)    // should succeed like now, but...
td.Cmp(t, a, b)    // should succeed, but...
td.Cmp(t, a, 1234) // should fail
```
why not.

**Comment by maxatome** [comment]
Or provide a pluggable mechanism (like we do with anchors) to allow a custom behavior for any type?

**Comment by stewi1014** [comment]
Yes, that's exactly what I mean.

I'm just wanting to compare the values inside, but it's possible someone may want to check that two reflect.Values are themselves the same. Perhaps they have a function returning a reflect.Value and they want to check that it's always addressable. Allowing custom behaviour for specific types could be quite powerful for this.

**Comment by maxatome** [comment]
Could you tell me if the branch https://github.com/maxatome/go-testdeep/tree/hooks fills your needs?

```go
func TestHooks(tt *testing.T) {
	got, expected := reflect.ValueOf(1234), reflect.ValueOf(1234)
	t := td.NewT(tt)
	t.AddCmpHook(func(got, expected reflect.Value) bool {
		return td.EqDeeply(got.Interface(), expected.Interface()) // or whatever test you want
	})
	t.Cmp(got, expected) // now succeeds
}
```
It is a quick proof of concept, it needs some docs + tests, but this feature could be available as is. What do you think about it?

`AddCmpHook` method only accepts a function matching the signature `func (T, T) bool` (it calls [this one](https://github.com/maxatome/go-testdeep/blob/hooks/internal/hooks/hooks.go#L31-L51) behind the scene).

**Comment by stewi1014** [comment]
Wow, that was fast!  I've imported it and am running some tests with it and it's working great, this definitely works. This library is revolutionising my testing :P

I did find something rather interesting. I noticed that CmpFalse in the test was failing. It took a while, but I figured out why. If I passed the number as a global variable to reflect.ValueOf it worked. Turns out the compiler is optimising the '1234' in ```got, expected := reflect.ValueOf(1234), reflect.ValueOf(1234)``` to the same address in memory on my machine.

e.g.

```go
func TestHooks(tt *testing.T) {
	got, expected := reflect.ValueOf(1234), reflect.ValueOf(1234)

	// get ptr field in got
	gotElemPtr := unsafe.Pointer((*(*[2]uintptr)(unsafe.Pointer(&got)))[1])

	// get ptr field in expected
	expectedElemPtr := unsafe.Pointer((*(*[2]uintptr)(unsafe.Pointer(&expected)))[1])

	// Same address!!
	fmt.Printf("%p %p\n", gotElemPtr, expectedElemPtr)

	// Just to sanity check it is actually the correct pointer.
	fmt.Println(*(*int)(gotElemPtr), *(*int)(expectedElemPtr))

	ttt := test.NewTestingTB(tt.Name())
	t := td.NewT(ttt)

	// Not the same reflect.Value or not-comparable
	td.CmpFalse(tt, func() bool {
		// A panic occurs when -tags safe:
		//   dark.GetInterface() does not handle private unsafe.Pointer kind
		defer func() { recover() }() //nolint: errcheck
		return t.Cmp(got, expected)
	}())

	// Add a hook to compare what is inside the reflect.Value
	t.AddCmpHook(func(got, expected reflect.Value) bool {
		return td.EqDeeply(got.Interface(), expected.Interface())
	})
	td.CmpTrue(tt, t.Cmp(got, expected))
}
```
I'm not sure how the compiler would handle
```go
n := 1234
got, expected := reflect.ValueOf(n), reflect.ValueOf(1234)
```
but I do know the compiler is guaranteed to not optimise global variables away irrelevant of their usage. It also appears as though this is non-deterministic, as clearing caches and trying again sometimes results in the test passing.

**Comment by maxatome** [comment]
Yes I experienced the same behavior with master branch of go during my tests. I will remove this `CmpFalse` test as nothing can guarantee it always succeeds (or fails) whatever the go version is used. Anyway, thanks for the explanations, when I saw that only master failed, I went to bed without digging any further...

We have to keep in mind that `reflect.Value` have not to be compared deeply, that is why `AddCmpHook` is useful :)

Note that the same applies for `reflect.Type`, it should be:
```go
t.AddCmpHook(func(got, expected reflect.Type) bool {
	return got == expected
})
```

I'll clean the code, write docs and some tests, and come back with a PR.

Do not forget to advertise go-testdeep! :D
